test/test_e2e_tts_fastspeech.py

```diff
@@ -0,0 +1,328 @@
+        decoder_concat_after=False,
+        transformer_init="pytorch",
+        initial_encoder_alpha=1.0,
+        initial_decoder_alpha=1.0,
+        reduction_factor=1,
+        loss_type="L1",
+        use_guided_attn_loss=False,
+        num_heads_applied_guided_attn=2,
+        num_layers_applied_guided_attn=2,
+        guided_attn_loss_sigma=0.4,
+        modules_applied_guided_attn=["encoder", "decoder", "encoder-decoder"]
+    )
+    defaults.update(kwargs)
+    return defaults
+
+
+def make_feedforward_transformer_args(**kwargs):
+    defaults = dict(
+        adim=32,
+        aheads=4,
+        elayers=2,
+        eunits=128,
+        dlayers=2,
+        dunits=128,
+        duration_predictor_layers=2,
+        duration_predictor_chans=128,
+        duration_predictor_kernel_size=3,
+        duration_predictor_dropout_rate=0.1,
+        positionwise_layer_type="linear",
+        positionwise_conv_kernel_size=1,
+        transformer_enc_dropout_rate=0.1,
+        transformer_enc_positional_dropout_rate=0.1,
+        transformer_enc_attn_dropout_rate=0.0,
+        transformer_dec_dropout_rate=0.1,
+        transformer_dec_positional_dropout_rate=0.1,
+        transformer_dec_attn_dropout_rate=0.3,
+        transformer_enc_dec_attn_dropout_rate=0.0,
+        use_masking=True,
+        use_scaled_pos_enc=True,
+        encoder_normalize_before=True,
+        decoder_normalize_before=True,
+        encoder_concat_after=False,
+        decoder_concat_after=False,
+        transformer_init="pytorch",
+        initial_encoder_alpha=1.0,
+        initial_decoder_alpha=1.0,
+        init_encoder_from_teacher=False,
+        init_encoder_module="all",
+        reduction_factor=1,
+        teacher_model=None,
+    )
+    defaults.update(kwargs)
+    return defaults
+
+
+@pytest.mark.parametrize(
+    "model_dict", [
+        ({}),
+        ({"use_masking": False}),
+        ({"use_scaled_pos_enc": False}),
+        ({"positionwise_layer_type": "conv1d", "positionwise_conv_kernel_size": 3}),
+        ({"encoder_normalize_before": False}),
+        ({"decoder_normalize_before": False}),
+        ({"encoder_normalize_before": False, "decoder_normalize_before": False}),
+        ({"encoder_concat_after": True}),
+        ({"decoder_concat_after": True}),
+        ({"encoder_concat_after": True, "decoder_concat_after": True}),
+    ])
+def test_fastspeech_trainable_and_decodable(model_dict):
+    # make args
+    idim, odim = 10, 25
+    teacher_model_args = make_transformer_args(**model_dict)
+    model_args = make_feedforward_transformer_args(**model_dict)
+
+    # setup batch
+    ilens = [10, 5]
+    olens = [20, 15]
+    batch = prepare_inputs(idim, odim, ilens, olens)
+
+    # define model
+    model = FeedForwardTransformer(idim, odim, Namespace(**model_args))
+    teacher_model = Transformer(idim, odim, Namespace(**teacher_model_args))
+    model.teacher = teacher_model
+    model.duration_calculator = DurationCalculator(model.teacher)
```

Comment:
Could we avoid accessing members directly? It'd be better to pass teacher to FeedForwardTransformer instead of manually injecting the value, otherwise we cannot get test coverage for `self._load_teacher_model(args.teacher_model)`  in FeedForwardTransformer.

Reply:
fixed.